### PR TITLE
`Logos`: Add temporary provider support with auto-discovery

### LINKED ISSUE #421: [Feature Proposal] `Logos` should support temp. providers
Logos should be able to support temp. hosted models, e.g., a Mac(Boook) hosts LLMs temp. via LMStudio and is reachable via a CloudFlare tunnel.

I want to quickly add it to Logos, so only specifying a URL and AUTH key. It then picks up automatically the available models. And when the Mac decides to go offline or stop hosting the models than Logos can gracefully handle it. (Should not even persist in the Database!)

### PR DESCRIPTION
**Pull request by wasnertobias**
## Summary

This PR adds **temporary (in-memory) LLM provider** support to Logos. It enables users to register ephemeral providers — such as LMStudio running on a local Mac exposed via a tunnel — which are auto-discovered, health-monitored, and seamlessly integrated into the existing proxy pipeline. All state lives in process memory and is **not persisted** to the database.

## Architecture

The feature is implemented as a self-contained `logos/src/logos/temp_providers/` module with three components:

| File | Responsibility |
|---|---|
| `registry.py` | Thread-safe, singleton, in-memory registry of `TempProvider` objects. Manages registration, lookup, health-status tracking, and stale-provider removal. |
| `discovery.py` | Auto-discovers models from a remote provider. Tries the **OpenAI-compatible** `GET /v1/models` endpoint first, then falls back to the **Ollama** `GET /api/tags` endpoint. |
| `health_monitor.py` | Background `asyncio` task that periodically pings every registered temp provider and auto-removes providers that stay unhealthy for longer than a configurable threshold (default 300 s). |

## Key Design Decisions

- **In-memory only** — temp providers are not written to the database, fulfilling the requirement for ephemeral/disposable registrations. Restarting Logos clears all temp providers.
- **Auth tokens (`tpk-*`)** — each registered temp provider receives a unique `tpk-<random>` token. Callers must include this token (via the `X-Temp-Provider-Token` header) when routing requests through the provider, preventing unauthorized access to someone else's temp provider.
- **Background health monitoring with auto-removal** — a `HealthMonitor` task runs every 30 s, checking each provider's liveness. Providers unreachable for >300 s are automatically purged.
- **Dual-protocol discovery** — `discover_models()` probes both OpenAI-compatible (`/v1/models`) and Ollama (`/api/tags`) APIs, so the user doesn't need to specify the server type.

## New Endpoints

| Method | Path | Description |
|---|---|---|
| `POST` | `/logosdb/add_temp_provider` | Register a temp provider. Body: `{ "url": "...", "name": "...", "auth_key": "..." }`. Returns provider info with discovered models and `auth_token`. |
| `DELETE` | `/logosdb/remove_temp_provider/{provider_id}` | Remove a temp provider. Owner or root authorization required. |
| `GET` | `/logosdb/temp_providers` | List temp providers with health status. Root sees all; non-root sees only own providers. |
| `POST` | `/logosdb/refresh_temp_provider/{provider_id}` | Force re-discovery of models on a temp provider. Owner or root authorization required. |

## Integration with the Proxy Pipeline

When `_execute_proxy_mode()` cannot find the requested model among the persisted deployments, it now falls back to the temp-provider registry. If a healthy temp provider serves the model (and the caller's `X-Temp-Provider-Token` matches), the request is forwarded directly to the provider's `/v1/<path>` endpoint via `_execute_temp_provider_request()`. If the provider is unreachable, it is marked unhealthy and a `503` is returned.

## Testing

**49 tests** across 4 test files:

| File | Tests | Coverage |
|---|---|---|
| `test_registry.py` | 22 | Registration, unregistration, listing, health marking, stale removal, singleton behavior, `find_provider_for_model` |
| `test_discovery.py` | 9 | OpenAI-compatible discovery, Ollama discovery, fallback logic, error handling |
| `test_health_monitor.py` | 7 | Health-check cycle, auto-removal, start/stop lifecycle |
| `test_temp_provider_endpoints.py` | 11 | All four endpoints (add, remove, list, refresh) including auth/authorization edge cases |

Run tests:
```bash
cd logos && python -m pytest tests/unit/temp_providers/ -v
```

## Summary by CodeRabbit

* **New Features**
  * Added support for temporary, in-memory LLM providers (e.g., local services via tunnel).
  * Automatic model discovery for OpenAI-compatible and Ollama endpoints.
  * New API endpoints to register, manage, list, and refresh temporary providers.
  * Health monitoring with automatic removal of stale providers.

### CODE REVIEW

File: logos/src/logos/temp_providers/registry.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    def __init__(self) -> None:
+        # Only initialise state once (singleton guard).
+        if not hasattr(self, "_providers"):
+            self._providers: Dict[str, TempProvider] = {}
+            self._lock = self._lock_cls()
+
+    def __new__(cls) -> "TempProviderRegistry":
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+        return cls._instance
```
[the remark this answers is not in the document]
Comment (wasnertobias) [reply logos/src/logos/temp_providers/registry.py]: Fixed in latest push. Added a class-level `threading.Lock()` with double-checked locking pattern in `__new__` to make singleton creation truly thread-safe.
